### ISSUE TITLE
Backward-compatible proposal to pass values to sub-charts by supporting value constraints

### ISSUE THREAD

**Issue by jorgecarleitao** [helm/helm#8576]
This is a proposal to extend Helm 3 in a backward compatible manner to support passing arbitrary values to subcharts without requiring developers to expose them on `values.yaml` or as globals. This is resolved as a consequence of the broader problem that this proposal addresses: add support to pass transformed/mapped values to `.Values`.

This proposal allows:

* chart developers to hint to chart users that some values are not intended to be overwriten, by removing them from `values.yaml`.
* pass values to sub-charts that are _derived_ from `values.yaml` (example below)
* improve chart's API by not asking users to use anchors or copy-pasting values passed to `helm create ...`.

In summary, this allows chart developers to have full discretion over the API that they want to expose in `values.yaml`, by hiding implementation detail currently on `values.yaml` to a different place.

# Specification

Introduce a new optional reserved file in a chart, `map.yaml`, on the root of the chart (same level as `Chart.yaml`) that:

1. is (in-memory) templated using all values passed via `values.yaml`, `--set`, etc.
2. is (in-memory) merged to `values.yaml` before all values are passed to `.Values`

when users call any `helm` commands that require rendering.

`map.yaml` uses the same syntax as any other template, and is expected to be rendered to a normal `yaml` file, like `values.yaml`. It's primary goal is to offer developers the opportunity to map/transform any value in `values.yaml` to a set of new values that are required by the chart.

For example, if `map.yaml` contains `sub.a: {{ print "%s-%s" .Values.b .Values.c }}`, `.Values.sub.a` equals `{b}-{c}` where `{b}` and `{c}` are the values passed by either `values.yaml` or `--set b= --set c=`.

When templating `map.yaml`, `.Values` is populated according to the same precedence rules of other templates (priority is given to values passed by the user over the ones in `values.yaml`).

A chart's `.Values.valueX` is populated according to the following precedence (higher takes priority):

1. `valueX` provided by the user (e.g. via `--set valueX`)
2. `valueX` rendered on `map.yaml` when that key is defined in `map.yaml` and `map.yaml` exists in the chart
3. `valueX` on `value.yaml`

This precedence applies equally to both a chart's main templates with subchart's templates. Specifically, subchart's that have a `map.yaml` use their `map.yaml` to populate their own `.Values` (together with whatever was passed by the parent chart).

# Motivation

Chart developers that use a subchart often specialize the subchart by requiring specific values to be passed to the subchart in order for the whole deployment to work as intended (a value constraint). Currently, chart developers have two ways to achieve this:

* add an entry `subchart: ...` to the chart's `values.yaml`
* refactor the subchart to use `globals` and use `globals` on `values.yaml`

These two approaches suffer from two distinct but fundamental limitations:

* `globals` are prone to naming collision across charts, both vertically (a chart with 5 sub-charts) and in depth (sub-chart of subchart of subchart).
* `subchart` requires the chart to expose all options to its user, even if some of them are not intended to be changed or break if changed. I.e. it does not allow the `subchart` to be an implementation detail of the parent chart.

The problem of passing values to sub-charts can be summarized as value constraints: a value passed to a sub-chart is often not independent, but derived from (constrained by) other values.

For simple charts, chart developers address value derivations through templates: create a named template that constructs a new value `Y` based on `.Values.X` and use the template throughout the chart. For subcharts, they do not have a mechanism to address this, and instead have to ask the chart's userto address it themselves (or use globals).

When a developer exposes a set of keys in `values.yaml` from subcharts that are related to each other, they are not doing so because the values are configurable, but because they have no mechanism to pass values to subcharts that are inter-dependent.

However, chart users rely on `values.yaml` to know how they can parameterise a chart. I.e. users only need to worry about those values and  appreciate when a chart's developer already though through which variables users need to worry about. This implies that keys in `values.yaml` that need some inter-dependency are an hassle to users, as they need to be educated (by chart developers) on which values to use and which constraints those values need to have.

Example of usage of this proposal: given a chart with

```yaml
# values.yaml
foo: ""

auth:
  app: "myapp"
```

```yaml
# map.yaml

# this chart expects that `bar` to be `test-{%foo}` (a constraint)
bar: {{ printf "test-%s" .Values.foo  }}

subchart:
  # same here: `some_app` is expected to be constructed in this specific way (a constraint)
  some_app: {{ printf "%s-%s" .Values.auth.app .Values.foo  }}
```

that a user installs with 

```bash
helm install ... --set foo="aa" --set auth.app="otherapp"
```

, the resulting `.Values` is

```yaml
foo: "aa"
bar: "test-aa"

auth:
  app: "otherapp"

subchart:
  some_app: "otherapp-aa"
```

This would be equivalent to have the chart written as

```
# values.yaml

foo: ""
bar: "test-"

auth:
  app: "myapp"

subchart:
   some_app: ""
```

and the user run

```
helm install ... --set foo="aa" --set bar="test-aa" --set auth.app="otherapp" --set "subchart.some_app=otherapp-aa"
```

Note how, in this call, the user needs to set `subchart.some_app` to be "%{auth.app}-%{foo}" for the chart to work as expected (a constraint). This is the core problem that this proposal addresses.

# Rationale

This design was selected primarily based on the following factors: 

* backward compatibility
* recursive support for sub-charts
* Clarity of chart's public API
* simplicity of usage

### backward compatibility

This design was considered due to its backward compatibility, and how it has a backward compatible migration path for chart developers. This change is backward compatible because:

1. `map.yaml` is optional and this functionality is only activated by its existence
2. ``--set subchart.valueX`` takes precedence over the values derived in `map.yaml` and thus charts can add `map.yaml` without breaking backward compatibility
3. `map.yaml` does not collide with existing templates (because it is not placed in `templates/`)

Furthermore, developers have a clear migration path from charts without `map.yaml` to charts with `map.yaml`. Specifically:

1. add `map.yaml` and to it any value that needs to be written in `.Values` that depend on any value from `values.yaml`
1. warn users in INFO that the chart plans to move to `map.yaml`, so that users no longer need to populate certain values.
1. remove entries from `values.yaml` that are now populated by `map.yaml`

They can decide when to move from one step to the next as part of their own chart's release life-cycle.

### Recursive support for sub-charts

Another reason for this design is that a sub-chart can itself have a `map.yaml`, and the precedence described in the specification works recursively. This allows charts to be self contained, as the developer of `subchart1` can decide which values to expose in `values.yaml` and which ones to derive using `map.yaml`, irrespectively of whether `subchart1` has `subchart2` or not. A chart that has subchart `subchart1` only needs to populate the values of `subchart1`'s `value.yaml`, and leave any value of `subchart2` to be populated by `subchart1`'s `map.yaml`.

### Clarity of chart's public API

So far, `values.yaml` has been a place to allow users to pass values to `.Values` and subcharts. A major reason for the choice of design in this proposal is that it effectively allows chart developers to make `values.yaml` _the public interface_ of their helm chart.

Even if a subchart would require a value `complicated_name_oauth_client_id` to be set, developers can expose that value as `oauth_client_id` and map it to `subchart.complicated_name_oauth_client_id` on `map.yaml`. Before, developers had to expose in `values.yaml` an implementation detail - that the chart uses a subchart, and users had to set `subchart.complicated_name_oauth_client_id` themselves.

### Simplicity of usage

The usage of the functionality hereby proposed is unchanged for users of helm charts: they continue to use helm exactly as before, and require no extra calls. This was a major reason for consideration of this design, as it ensures that no training is required for chart users.

For chart developers, the functionality here proposed is easy to explain: `values.yaml` contains all values that a user can change independently. All dependent (constrained, derived) values are placed in `map.yaml` so that they are available in `.Values`. For backward compatibility, keep placing those values also in `values.yaml`, so that they can continue to be passed by users using an Helm version that does not support `map.yaml`.

Another simplification of this design is that developers no longer need to write named templates to construct names derived from `.Values`. They can just write them in `map.yaml`. An example would be `componentName: {{ .Release.name }}-component` on `map.yaml`, thus allowing them to use `.Values.componentName` on the chart (instead of declaring the named template to the same effect).

This proposal considers `map.yaml` a template for two reasons:

1. because chart developers are already familiar with templates;
2. because templates provide full flexibility on the transformations they support

# Backwards Compatibility

This proposal is backward compatible.

# Reference Implementation

A proof of concept is presented in [#8580](https://github.com/helm/helm/pull/8580).

**Comment by mattfarina** [comment]
Thank you for the proposal.

I would like to back up to the problem for a moment. The two roles involved in this are the chart developer and the Helm user who is installing the chart. These two roles may not be the same person and they may not be at the same organization. The Helm user may not even be an expert Kubernetes user or something who has any expertise in the app they are installing. This also affects the experience for that Helm user.

What is the problem in the current experience you are trying to solve? I'm not sure I'm entirely clear. Is it that you want to have a parent chart set sub-chart properties in a way that the Helm user cannot override? If so, why is that?

The motivation and impact on users is very important to us. We want to fully understand it.1

It also occurred to me that schemas could be used to enforce certain values being passed to the subchart. Changes to values would then fail validation.

**Comment by jorgecarleitao** [comment]
Thanks a lot, @mattfarina for taking the time to look at this. Maybe I can use a concrete example to show case this.

Let's say that I would like to offer an helm chart that deploys `jupyterhub` together with some other application, that both will rely on Azure AD for authentication and authorization, and that they are both deployed on the same subdomain, `example.com`. The canonical chart for `jupyterhub` is this one: https://jupyterhub.github.io/helm-chart/ . To configure it to use Azure AD, we need to pass

```
auth:
  type: azuread
  azuread:
    clientId: ""
    clientSecret: ""
    tenantId: ""
    callbackUrl: ""
```

to it.

So, as a developer of a chart that uses this subchart,

```
dependencies:
  - name: jupyterhub
    repository: https://jupyterhub.github.io/helm-chart/
    version: 0.9.0
```

I need to add to `values.yaml` the above configuration. Now, my other application has a more generic configuration, and only requires an openid-configuration, that on azure looks like this: `https://login.microsoftonline.com/$TENANT_ID/v2.0/.well-known/openid-configuration`

So, my `values.yaml` is now:

```
app:
    confUrl: ""
    clientId: ""
    clientSecret: ""
    subdomain: ""

jupyterhub:
  auth:
    type: azuread
    azuread:
      clientId: ""
      clientSecret: ""
      tenantId: ""
      callbackUrl: ""
```

This `values.yaml` suggests that users of my chart need to specify 7 parameters, but this chart is not really parameterized by 7 parameters, since `jupyterhub.azuread.tenantId` and `app.confUrl` are both constructed from the `TENANT_ID` (what I call a derived value). Currently, my chart's documentation would have to state

> the tenant that you pass to `app.confUrl` must be `https://login.microsoftonline.com/$TENANT_ID/v2.0/.well-known/openid-configuration`, where `$TENANT_ID` must be equal to the value passed to `jupyterhub.auth.azuread.tenantId`
>
> Furthermore, the `callbackUrl` must be on the same domain as `app.subdomain`.
> Furthermore, the `app.clientSecret` must be on the same as `jupyterhub.auth.azuread.clientSecret`
> ...

Having to write this on a README is not really needed, and is a just a consequence of `values.yaml` being both the interface to users of charts and the "place to pass values to subcharts". More importantly, it is an hassle both to users and developers, and lead to an overall poor experience and challenges in re-using charts as subcharts.

This proposal overcomes this by offering chart developers the option to not have to write that. In this example, they would write instead:

```
# values.yaml

tenantId: ""
subdomain: ""
clientSecret: ""
clientId: ""
```

```
# map.yaml
app:
    confUrl: {{ printf "https://login.microsoftonline.com/%s/v2.0/.well-known/openid-configuration" .Values.tenantId }}
    clientId: {{ .Values.clientId }}
    clientSecret: {{ .Values.clientSecret }}
    subdomain: {{ .Values.subdomain }}

jupyterhub:
  auth:
    type: azuread
    azuread:
      clientId: {{ .Values.clientId }}
      clientSecret: {{ .Values.clientSecret }}
      tenantId: {{ .Values.tenantId }}
      callbackUrl: {{ printf "https://notebook.%s/callback" .Values.subdomain }}
```

`map.yaml` is an implementation detail of the chart that chart users can't do anything to it. `values.yaml` is the public interface of the chart, and users can overwrite values as before.

We dramatically simplified the usage and documentation of this chart with this small trick. This is the benefit of this proposal and associated draft PR.

**Comment by paulczar** [comment]
This looks to me like something that [helmfile](https://github.com/roboll/helmfile) solves very well by letting you template out the values.yml passed to charts and would allow you to do exactly what you're doing in `map.yaml` above. see for an example this [grafana values file](https://github.com/paulczar/platform-operations-on-kubernetes/blob/master/charts/grafana/helmfile/values.yaml.gotmpl) that helmfile is templating with a combination of env vars and Values.

I realize that would involve adding an extra tool to your process, but it does [If I understand it right] solve this problem quite neatly.

**Comment by jorgecarleitao** [comment]
@paulczar, Yes! This is definitely (one of the) problems that an helmfile solves. But, like you mentioned, it requires yet another tool. I am not arguing that there are no options around to address this issue. But the fact that there are such options already hints that this is a significant hassle (or people would not bother to address it).

The issue is that as a chart developer, I need to _ask_ my consumers to do something (pass two values, with some bash script, use helmfile, etc) when composing charts, IMO due to a avoidable limitation of helm.

The offer here is a mechanism to offer a chart with subcharts without requiring the chart's consumer to do something. IMO `helmfile` is asking the consumer more than some bash script populate the `values.yaml` to guarantee the required values' invariants. For example, in secure environments, it requires an extra audit to a binary that has permissions to install other binaries on an infrastructure.

We sell Helm as _the package manager for Kubernetes_. We should not offer "but you need to use an external tool" as the answer to subcharts effectively.

**Comment by TBBle** [comment]
This appears to be a subset of the problem that #6876 would solve, and specifically for sub-charts.

As a first thought, if we're doing this specifically for sub-charts, should this be more-tied to the dependencies list? This seems like a wider-functioning version of `export-values` (#3242) so perhaps rather than a hard-coded single `map.yaml`, either put the mappings in the Chart.yaml like `export-values`, or have the mapping file be _named_ in Chart.yaml per sub-chart. Or, like #6876, a new subdirectory named 'mappings' (or something) with files matching the sub-chart alias name.

I'd definitely like to see either something like this, or #6876 land, since we build a super-chart and a bunch of sub-charts, so more-and-more of our configuration ends up in `global`.

Personally, I'd prefer #6876 because it gives us more power to simplify values.yaml even in a single chart, while it seems for this approach it's not expected that map.yaml could affect the values seen by the current chart, only sub-charts.

**Comment by jorgecarleitao** [comment]
@TBBle , the `map.yaml` proposed here would use all values from `Values.yaml` to create a new `values` passed to both the current chart and all sub-charts. In particular, this also helps for a single chart: if `Values.yaml` contains

```
foo: ""
```

and `map.yaml` contains

```
bar: {{ printf "test-%s" .Values.foo | quote }}
```

the values passed to all templates in the chart when templated with `--set foo="aaa"` is

```
foo: "aaa"
bar: "test-aaa"
```

**Comment by TBBle** [comment]
#6876 should be backwards compatible, I don't see what the behaviour breakage is, except that it fixed one problem, where disabled sub-charts are read inappropriately.

As far as I'm reading the documentation of #6876 , and based on your description of _map.yaml_ here, the difference is that #6876 gives us a whole directory templated-values-files like _map.yaml_. I also like that, because it gives more flexibility to chart creation structure, like having a _templates/_ directory rather than a _template.yaml_, but it's not a deal-breaker for me.

Perhaps I am misunderstanding something. If this PR was changed to replace _map.yaml_ with _valuemaps/*.yaml_, what is the behaviour difference between this and #6876?

Conceptually, the simplest approach I can see is if the feature (however it's actually implemented) is in-effect just generating files to append `--values` to the install/template call. So maybe it's worth identifying the ways in which this or #6876 is _not_ doing that.

The main difference I can glean in terms of behaviour is that this code will override the templates-values-files with any `--set` values, while #6876 takes the `--set` values for rendering the templates-values-files, but does not then override the resulting values _again_ with `--set` values.

Or in terms of a 'generated command line', this PR does:
```
helm template-values --values user-provided-values.yaml --set user-provided-set
helm ... --values user-provided-values.yaml --values values-generated-from-map.yaml --set user-provided-set
```
and #6876 is
```
helm template-values --values user-provided-values.yaml --set user-provided-set
helm ... --values user-provided-values.yaml --values values-generated-from-map.yaml
```

If that's the accurate difference, I prefer the behaviour of this one as a chart user, since I can always workaround issues in the chart with --set, and as a chart developer, I only lose the ability to _definitely_ prevent users from breaking the chart assumptions by overriding things I'm doing in the generated values. And I don't have that ability right now anyway, so it's clearly not important.

**Comment by jorgecarleitao** [comment]
I agree with you that one difference is the `valuemaps/*` vs a single `map.yaml`. I do not understand how multiple map files in #6876 helps, though: `.Values` is a HashMap with unique keys, which requires the developer to pass 1, and only 1, value to a given key. With multiple files, `valuemaps/*`, the developer needs to know that there is no intersection between any key on any of the files in `valuemaps/*`. This is exactly what a single `yaml` file does, as keys are unique and most `yaml` linters will warn about duplicates.

The main difference between the two is that this PR does not require a second call of `helm`:

```
helm ... --set ...
```

will template the `map.yaml` in memory and directly merge them to `Values.yaml` (in memory) when templating.

This means that, as a user, I _do not care_ whether the chart has a `valuemaps/*` or `map.yaml` or whatever: I make the same call as I was doing before. With #6876, I need to *know* that the `chart` uses `valuemaps`, and use that information to decide if I should template its valuesmaps or not.

In other words, #6876 is backward incompatible because it requires an extra call of `helm`. I.e. if I had a chart deployed and helm 3, I upgrade to helm 3.X with #6876 on it, also bump the chart to a version that supports #6876, the whole deployment fails because now I need to make the extra call.

Finally, #6876 continues to have an issue with a sub-chart of a sub-chart, as the sub-chart with `valuemaps/*` now requires an extra call of `helm template-values`, and thus continues to be difficult to scaffold a sub-chart on a chart: I will need to tell my user to template `sub-chart`'s `valuesmaps`, then template my chart's `valuesmaps`, and then template the chart.

The offer in this proposal is to leave all those details to each of the sub-chart developers.

Does this make sense?

**Comment by TBBle** [comment]
Maybe I wasn't clear earlier: Neither approach _actually_ requires the user to call Helm twice.

I was trying to illustrate the difference in the values applied to the templates/ directory. Specifically, the difference in the application of `--set` compared to the generated values. So I'll try again.

#6876 is roughly "Make values.yaml a template", this change is roughly "Generate an extra --values parameter from a template".

As I said before, I think I like this approach more. It's a _larger_ change, but more flexible for the user, at the cost of diverging the behaviour of `--set` and `--values` further, since the author of _map.yaml_ _can_ override things from `--values`, but _cannot_ override things from `--set`. In #6876, the chart author can override both `--values` and `--set` as they are not distinguished, and `--set` is not reapplied after the _map.yaml_ is processed.

----

Helm *already* supports passing `--values` multiple times, they are applied in order, overlapping keys are merged. The _resulting_ `.Values` is a hashmap with one value per key, but you can build it from multiple YAML files, as well as the `--set` command-line parameters on top. (This actually introduces some issues with arrays, as overlaying them loses existing elements, but that's a different issue, e.g. #5711)

So I can have _values/00_set_some_defaults.yaml_, and then also have _values/01_tls_is_enabled.yaml_, where the latter is wrapped in one big `{{ - if .Values.tls -}}` and then overrides a smattering of things that were set in the previous file, such as `subchartA.ingress.annotations`, `subchartB.ingress.annotations`, etc. Having to do that in a single file requires me to have a huge YAML file with lots of `{{ - if .Values.tls - }}` ... `{{ else }}` which is harder to write, _and_ harder to maintain.

**Comment by jorgecarleitao** [comment]
@TBBle , now I see what you mean, sorry for the misunderstanding and thank your for the pacience. I get it and I agree with you. 👍 

The #8580 does not reapply yet, but it should, to give a backward compatible migration path of the charts: when developers add `map.yaml`, those values should not overwrite the values from `--set` on the final `.Values` to not break backward compatibility of the charts. This also offers chart users full flexibility. I did not bother doing so yet as it depends on whether this gains traction or not.

**Comment by TBBle** [comment]
Since we're talking about a new feature, there's not really a backward-compatibility concern however it's implemented. Charts are SemVer'd, so if a chart author adds a _map.yaml_, they're probably going to need to bump the chart major version anyway, since that _map.yaml_ may overwrite values passed in via `--values`, and is probably going to trade-off by simplifying the chart's _values.yaml_ as well, so the user will have to change their existing `--values` and `--set` inputs. That's a chart change, though, not a Helm change.

Unchanged charts will have unchanged behaviour, so it remains backwards-compatible. (Unless they happened to have a map.yaml for unrelated reasons, but that will only affect the chart packager, not the chart user)

The reason I see this approach as a bigger change is that by reapplying `--set`, we end up with a newly-different behaviour between `--values` and `--set` (whether _map.yaml_ can overwrite the value), where currently the only differences between those two are the "retain" behaviour (which I personally find inconsistent, and actively avoid using) and the array-merging behaviour (which is already a rich source of GitHub issues >_<).

The analogy of this approach as "appends another `--values` parameter" makes that a _consistent_ difference (`--set` already overrides `--values` irrespective of command-line order), with the only weirdness being that the "appended `--values`" was generated with full context of the existing values, both `--values` and `--set`. That's a problem with the analogy though, not the behaviour.

This analogy also makes the 'support multiple files' request have an obvious, consistent merging behaviour, since it would be the same way `--values` works.

**Comment by jorgecarleitao** [comment]
I agree with your point about backward-compatibility. My argument was more that we should not only worry about backward compatibility with helm itself, but also think in terms of the charts themselves, as their developers are our core audience.

They would most likely like to introduce `map.yaml` without having to maintain another version of the chart for older versions of Helm. In this direction, I was thinking that we should reapply both `--set` and `--values`, i.e. in the specification above, we should have

1. `valueX` provided by the user (via `--set valueX` or `--values` with a `valueX`)
2. `valueX` rendered on `map.yaml` when that key is defined in `map.yaml` and `map.yaml` exists in the chart
3. `valueX` on `value.yaml`

However, if I understood correctly, your suggestion is that the specification should be

1. `valueX` rendered on `map.yaml` when that key is defined in `map.yaml` and `map.yaml` exists in the chart
2. `valueX` provided by the user (e.g. via `--set valueX`)
3. `valueX` on `value.yaml`

(1 and 2 swapped)

so that `map.yaml` behaves like `--values`. I would be fine with it also, but IMO will significantly impact adoption as it forbids developers from having a chart that supports both versions of Helm. OTOH, it makes it clearer that certain `.Values` are set by `map.yaml` irrespectively of what the user passes to them.

**Comment by TBBle** [comment]
It's only possible to write a chart that uses map.yaml but also functions compatibly without it if you're only using map.yaml for cosmetic restructuring of inputs. It's not really possible if you're using it to implement things that are not reasonable without it, such as replacing a bunch of `tpl` calls or `.Values.global` data used to co-ordinate unrelated sub-charts.

To be backwards-compatible with Helm versions, I'd end up maintaining two versions of my chart anyway, just all in-place using `if` rather than as two distinct charts. For my main use case (which is private charts for known users), I'd just be setting the minimum Helm version to whatever version implements this feature, if I'm using the feature.

If you reapply _both_ `--set` and `--values`, then you're in-effect "prepends a --values parameter".

I'm still somewhat unsure if I prefer that, or the second suggestion of "pass values through map.yaml and then merge that on top of the existing values" (which is how #6876 behaves, I believe).

Side-note: I'm not sure that 'map.yaml' is the right name for the file. It would make sense if all this feature did was remap values to other values (a bit like import-values as I mentioned earlier), but since it's being rendered by the template engine, it's capable of doing a lot more. What about _valuesTemplate.yaml_?

**Comment by jorgecarleitao** [comment]
Those are good points. I would try to gather more feedback, and if others also feel the same way, let's change the priority order.

To be specific, your proposal is that the priority is:

1. `valueX` rendered on `map.yaml` when that key is defined in `map.yaml` and `map.yaml` exists in the chart
2. `valueX` provided by the user via `--set valueX`
3. `valueX` on `value.yaml`

In item 2,

> via `--set valueX`

or

> via `--set valueX` or `--values` with `valueX`" ?

I also agree with you wrt to the `map.yaml -> valuesTemplate.yaml`.

**Comment by TBBle** [comment]
For that ordering, the latter. If_valuesTemplate.yaml_ is going to be authoritative when it's merged onto the existing calculated values, then the existing calculated values should remain the same for both `--set` _and_ `--values`. That's how both #8580 and #6876 work right now, anyway, and it seems least-surprising.

In this flow, using [template functions](https://masterminds.github.io/sprig/dicts.html) as pseudocode, we get:
```
$values1 := mergeOverwrite values.yaml `--values A.yaml` `--values B.yaml` `--set X` `--set Y` globals import-values
$generatedValues := tpl valueTemplate.yaml $values1
$values2 := mergeOverwrite $values1 $generatedValues
```
and then render the templates using `$values2`.

The other option is effectively what I've described as inserting a values.yaml
```
$values1 := mergeOverwrite values.yaml `--values A.yaml` `--values B.yaml` `--set X` `--set Y` globals import-values
$generatedValues := tpl valueTemplate.yaml $values1
$values2 := mergeOverwrite values.yaml $generatedValues `--values A.yaml` `--values B.yaml` `--set X` `--set Y` globals import-values
```

`$values1` here is roughly what Helm does now, although I'm not sure about the `global`s/`import-values` interaction. I know off-hand you can `import-values` `global`s from a sub-chart, since I've had to do it in code here, to carry data from one chart to a sibling chart via a parent chart. `globals` and `import-values` are complicated because they depend on which chart you're evaluating, _and_ the evaluation of the sub-chart. Most of the refactoring in #6876 was to make that work.
